src/modules/battery_resistance_est/battery_resistance_est.cpp

```diff
@@ -0,0 +1,255 @@
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "battery_resistance_est.hpp"
+
+InternalRes::InternalRes() :
+	ModuleParams(nullptr),
+	WorkItem(MODULE_NAME, px4::wq_configurations::lp_default)
+{
+}
+
+bool InternalRes::init()
+{
+	if (!_battery_sub.registerCallback()) {
+		PX4_ERR("battery status callback registration failed!");
+		return false;
+	}
+
+	_param_est(0) = r_s;
+	_param_est(1) = (r_t + r_s)/(r_t*c_t);
+	_param_est(2) = 1.0f/(r_t*c_t);
+	_param_est(3) = v_oc/(r_t*c_t);
+
+	_adaptation_gain(0) = 0.0001f;
+	_adaptation_gain(1) = 0.0001f;
+	_adaptation_gain(2) = 0.0001f;
+	_adaptation_gain(3) = 0.0001f;
+
+	return true;
+}
+
+float InternalRes::extract_parameters() {
+
+	const float _r_steady_state = _param_est(0);
+	const float _r_transient = (_param_est(1)/_param_est(2)) - _param_est(0);
+	const float _voltage_open_circuit = _param_est(3)/_param_est(2);
+
+	//calculate bat1_r_internal
+	const float _internal_resistance = (_voltage_filtered_v - _voltage_open_circuit) / (-_current_filtered_a);
+
+	//logging
+	inter_res.r_transient = _r_transient;
+	inter_res.r_steady_state = _r_steady_state;
+	inter_res.voltage_open_circuit = _voltage_open_circuit;
+	inter_res.bat1_r_internal = _internal_resistance;
+
+	return _internal_resistance;
+
+}
+
+
+void InternalRes::write_internal_resistance(const float _voltage_estimation_error, const float _internal_resistance){
+
+	//store best esimate
+	if((abs(_voltage_estimation_error)) <= abs(best_prediction_error))
+	{
+		best_prediction = _internal_resistance;
+		best_prediction_error = _voltage_estimation_error;
+	} else if (best_prediction_error_reset){
+		best_prediction = _internal_resistance;
+		best_prediction_error = _voltage_estimation_error;
+		best_prediction_error_reset = false;
+	}
+
+	const float time_since_param_write = (hrt_absolute_time()- last_param_write_time)/ 1e6f;
+
+	// save bat1_r_internal only periodically
+	if (time_since_param_write >= _inter_res_update_period.get()){
+
+		//BAT${i}_R_INTERNAL  increment: 0.01
+		best_prediction = round(best_prediction*100)/100;
+
+		//clamp BAT${i}_R_INTERNAL min: -1.0  max: 0.2
+		if (best_prediction > 0.2f){
+	    		best_prediction = 0.2f;
+		} else if (best_prediction < -1.0f){
+	    		best_prediction = -1.0f;
+		}
+
+		//Set px4 Internal resistance using simplified Rint Model
+		_bat1_r_internal.set(best_prediction);
+		_bat1_r_internal.commit();
```

Comment:
We can't have regular param notifications & savings. Rather use uORB for updates to other modules, and save the param on disarming.

Reply:
@bkueng Thanks for your feedback. Sorry, I am not aware of the issues this could cause. As suggested, I made some changes so that r_internal in lib/battery/battery.cpp is updated via uorb. 

What would the purpose of saving the param on disarm be?

Reply:
Just if you want to have it persistent across reboots, e.g. for faster initialization.

Reply:
@bkueng makes sense. I've made the required changes. Thanks!